zerver/actions/users.py

```diff
@@ -472,6 +472,10 @@ def do_change_user_role(
         },
     )
     maybe_enqueue_audit_log_upload(user_profile.realm)
+    if settings.BILLING_ENABLED and UserProfile.ROLE_GUEST in [old_value, value]:
+        billing_session = RealmBillingSession(user=user_profile, realm=user_profile.realm)
+        billing_session.update_license_ledger_if_needed(timezone_now())
```

Comment:
I wonder if it'd make sense to instead make this part of `maybe_enqueue_audit_log_upload`? As I understand it, that's the function we call to make sure billing data is correct for self-hosted systems in the same set of scenarios, and it might be a nice cleanup to just have a single function that code paths use to handle both.

I suppose we might need a different name, like `update_billing_records_if_needed(realm)`, and maybe that'd be inside a check for either role being a guest here.

Probably that's enough of a change that it can be its own follow-up PR if you like the idea.